test/Validate.SSA.cpp

```diff
@@ -0,0 +1,699 @@
+          OpDecorate %missing Restrict
+%voidt  = OpTypeVoid
+%intt   = OpTypeInt 32 1
+%ptrt   = OpTypePointer UniformConstant %intt
+%var    = OpVariable %ptrt UniformConstant
+%2      = OpTypeFunction %voidt
+%3      = OpFunction %voidt None %2
+%4      = OpLabel
+          OpReturn
+          OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardMemberDecorateGood) {
+  char str[] = R"(
+          OpCapability Matrix
+          OpMemberDecorate %struct 1 RowMajor
+%intt   = OpTypeInt 32 1
+%vec3   = OpTypeVector %intt 3
+%mat33  = OpTypeMatrix %vec3 3
+%struct = OpTypeStruct %intt %mat33
+)";
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, ForwardMemberDecorateInvalidIdBad) {
+  char str[] = R"(
+          OpCapability Matrix
+          OpMemberDecorate %missing 1 RowMajor
+%intt   = OpTypeInt 32 1
+%vec3   = OpTypeVector %intt 3
+%mat33  = OpTypeMatrix %vec3 3
+%struct = OpTypeStruct %intt %mat33
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardGroupDecorateGood) {
+  char str[] = R"(
+          OpCapability Matrix
+          OpDecorate %dgrp RowMajor
+%dgrp   = OpDecorationGroup
+          OpGroupDecorate %dgrp %mat33 %mat44
+%intt   = OpTypeInt 32 1
+%vec3   = OpTypeVector %intt 3
+%vec4   = OpTypeVector %intt 4
+%mat33  = OpTypeMatrix %vec3 3
+%mat44  = OpTypeMatrix %vec4 4
+)";
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, ForwardGroupDecorateMissingIdBad) {
+  char str[] = R"(
+          OpCapability Matrix
+          OpDecorate %dgrp RowMajor
+%dgrp   = OpDecorationGroup
+          OpGroupDecorate %dgrp %missing %mat44
+%intt   = OpTypeInt 32 1
+%vec3   = OpTypeVector %intt 3
+%vec4   = OpTypeVector %intt 4
+%mat33  = OpTypeMatrix %vec3 3
+%mat44  = OpTypeMatrix %vec4 4
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardGroupDecorateDecorationGroupDominateBad) {
+  char str[] = R"(
+          OpCapability Matrix
+          OpDecorate %dgrp RowMajor
+          OpGroupDecorate %dgrp %mat33 %mat44
+%dgrp   = OpDecorationGroup
+%intt   = OpTypeInt 32 1
+%vec3   = OpTypeVector %intt 3
+%vec4   = OpTypeVector %intt 4
+%mat33  = OpTypeMatrix %vec3 3
+%mat44  = OpTypeMatrix %vec4 4
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardDecorateInvalidIdBad) {
+  char str[] = R"(
+          OpMemoryModel Logical GLSL450
+          OpEntryPoint GLCompute %3 ""
+          OpExecutionMode %3 LocalSize 1 1 1
+          OpDecorate %missing Restrict
+%voidt  = OpTypeVoid
+%intt   = OpTypeInt 32 1
+%ptrt   = OpTypePointer UniformConstant %intt
+%var    = OpVariable %ptrt UniformConstant
+%2      = OpTypeFunction %voidt
+%3      = OpFunction %voidt None %2
+%4      = OpLabel
+          OpReturn
+          OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardFunctionCallGood) {
```

Comment:
optional: could simplify this by using void(void) functions.
Don't need the int types or int constant values.